### ISSUE TITLE
MatchSpectra function

### ISSUE THREAD

**Issue by tiwa1125** [rformassspectrometry/MetaboAnnotation#52]
Hi, 
I tried to use the MetaboAnnotation to annotate my metabolites, and I used the HMDB to get the match, while when I used the code below, I even did not get any matches, that is weird, as I am sure there are many metabolites could be identified. Do you know what is the problem?
> data("hmdb", package = "SpectraTutorials")
> mtches <- matchSpectra(pest_ms2, hmdb,
+                        param = CompareSpectraParam(requirePrecursor = TRUE,
+                                                    ppm = 30))
> mtches
Object of class MatchedSpectra 
Total number of matches: 0 
Number of query objects: 5220 (0 matched)
Number of target objects: 458963 (0 matched)

Best 
Tingting

**Comment by jorainer** [comment]
I assume the `pest_ms2` contains MS2 spectra from the *pesticide* test data? If that's the case you'll not find matches in HMDB because HMDB provides human metabolite data and does not contain (AFAIK) plant pesticides You should however get some matches if you use MassBank as that resource contains also MS2 spectra from not naturally occuring compounds.

**Comment by tiwa1125** [comment]
Hi Johannes,
Thanks for your reply, I did not use the test data, but my own data, it is a sample from human blood, I run it on bruker qtof with DDA mode, so I assume I should get some matches, since I already know many metabolites are there by matching with the authentic standards. Did I make something wrong?
> pest_ms2
MSn data (Spectra) with 5220 spectra in a MsBackendMzR backend:
       msLevel     rtime scanIndex
     <integer> <numeric> <integer>
1            2    15.388        17
2            2    15.488        18
3            2    15.588        19
4            2    15.688        20
5            2    15.788        21
...        ...       ...       ...
5216         2   600.354      5753
5217         2   600.454      5754
5218         2   600.554      5755
5219         2   600.654      5756
5220         2   600.764      5757
... 34 more variables/columns.

file(s):
TX41855.mzML
Processing:
Filter: select MS level(s) 2 [Fri Dec 31 17:42:26 2021]

Best
Tingting

**Comment by jorainer** [comment]
I would maybe suggest to remove some of the low intensity peaks in your MS2 spectra (similar to what I did in the [tutorial](https://jorainer.github.io/SpectraTutorials/articles/Spectra-matching-with-MetaboAnnotation.html)). Then, importantly, HMDB does not provide precursor m/z values, so you should use `requirePrecursor = FALSE`. Unfortunately, this will have a huge impact on performance because then you'll compare all your experimental spectra with all database spectra. Maybe it would also be helpful to subset the HMDB data excluding predicted spectra (using `hmdb <- hmdb[!hmdb$predicted]`)?

If you still miss some of the obvious hits, it might be that you have to further increase the `ppm` or `tolerance`. Some of the spectra in HMDB seem to be mis-calibrated (I had to use a large tolerance value to find some of our pure standards in HMDB data).

**Comment by tiwa1125** [comment]
Hi,
I finally made it work, while I got some weird results, I think the target compound ID does not fit with the fragments it found, did I do something wrong?
rtime
precursorMz
target_precursorMz
score
target_compound_id
target_predicted
basePeakMZ
target_spectrum_id
23.266
115.0373611
NA
0.855602
HMDB0000002
FALSE
0
2512
23.266
115.0373611
NA
0.855225
HMDB0000002
FALSE
0
445976
80.006
218.140564
NA
0.84175
HMDB0000019
FALSE
0
31
81.226
218.1393585
NA
0.774926
HMDB0000019
FALSE
0
31
112.683
232.1546326
NA
0.831935
HMDB0000019
FALSE
0
31
160.654
246.1705627
NA
0.750262
HMDB0000019
FALSE
0
31
247.12
304.2121582
NA
0.747934
HMDB0000019
FALSE
0
31
248.145
304.2121277
NA
0.77728
HMDB0000019
FALSE
0
31
293.157
286.2027588
NA
0.839669
HMDB0000019
FALSE
0
31
320.075
288.2186584
NA
0.751669
HMDB0000019
FALSE
0
31

Best
Tingting

**Comment by jorainer** [comment]
Note that the masses of the compounds are exact masses, not m/z values, thus, they might indeed be smaller than the precursor ion, depending on what ion/adduct was fragmented. Hm, and you're right, the compound name is not available in that data.

I would thus suggest to do something different:

- Install the `CompoundDb` package with `BiocManager::install("RforMassSpectrometry/CompoundDb")`.
- Download the CompDb database for HMDB (5.0 from 2021-11-02) from [here](https://github.com/jorainer/MetaboAnnotationTutorials/releases/tag/2021-11-02) (i.e. download the .sqlite file to your computer).
- Use that database instead of the `hmdb` variable you use above.

Basically, in your script, replacing 

```r
data("hmdb", package = "SpectraTutorials")
```

with

```r
library(CompoundDb)
cdb <- CompDb(<sqlite file name>)
hmdb <- Spectra(cdb)
```

(`<sqlite file name>` would then be the path and filename of the downloaded sqlite file) should allow you to use this resource instead. You could also have a look at this new tutorial for the [MetaboAnnotation](https://jorainer.github.io/MetaboAnnotationTutorials/) package for more information.
